src/Samples/SubModelSeq/Program.fs

```diff
@@ -13,177 +29,192 @@ module List =
 
-  let asLeaf a =
-    { Data = a
-      Children = [] }
+    let rec map f t =
+      t |> f |> mapChildren (f |> map |> List.map)
 
-  let dataOfChildren n =
-    n.Children |> List.map (fun nn -> nn.Data)
+    let asLeaf a =
+      { Data = a
+        Children = [] }
 
-  let rec map f n =
-    let nn = n |> f
-    { nn with Children = nn.Children |> List.map (map f) }
+    let addChild a = a |> asLeaf |> List.cons |> mapChildren
+    
+    let dataOfChildren n =
+      n.Children |> List.map getData
+    
+    let rec preorderFlatten n =
+      n :: List.collect preorderFlatten n.Children
 
-  let rec mapData f n =
-    { Data = n.Data |> f
-      Children = n.Children |> List.map (mapData f) }
-
-  let rec preorderFlatten n =
-    n :: List.collect preorderFlatten n.Children
 
 module App =
 
   type Model =
     { SomeGlobalState: bool
-      DummyRoot: Tree.Node<Counter> }
+      DummyRoot: RoseTree<Identifiable<Counter>> }
 
-  let asDummyRoot ns : Tree.Node<Counter> =
-    { Data = Counter.create() // Placeholder data to satisfy type system. User never sees this.
-      Children = ns }
+  type Msg =
+    | ToggleGlobalState
+    | AddCounter of parent: Guid
+    | CounterMsg of Guid * CounterMsg
+    | Remove of Guid
+    | MoveUp of Guid
+    | MoveDown of Guid
+
+
+  module SomeGlobalState =
+    let get m = m.SomeGlobalState
+    let set v m = { m with SomeGlobalState = v }
+    let map f = f |> map get set
+  module DummyRoot =
+    let get m = m.DummyRoot
+    let set v m = { m with DummyRoot = v }
+    let map f = f |> map get set
+
+  let createIdentifiableCounter () =
+    { Id = Guid.NewGuid ()
+      Value = Counter.init }
 
   let init () =
     { SomeGlobalState = false
-      DummyRoot = [ Counter.create() |> Tree.asLeaf ] |> asDummyRoot }
+      DummyRoot =
+        createIdentifiableCounter () // Placeholder data to satisfy type system. User never sees this.
+        |> RoseTree.asLeaf
+        |> RoseTree.addChild (createIdentifiableCounter ()) }
+
+  let hasId id n = n.Data.Id = id
+
+  let addChildCounter nId =
+    { Id = nId
+      Value = Counter.init }
+    |> RoseTree.addChild
+
+  let swapCounters swap nId =
+    nId
+    |> hasId
+    |> List.tryFindIndex
+    >> Option.map swap
+    >> Option.defaultValue id
+    |> Func.dedup
+    |> RoseTree.mapChildren
+
+  let update = function
+    | ToggleGlobalState -> not |> SomeGlobalState.map
+    | CounterMsg (nId, msg) -> msg |> Counter.update |> Identifiable.mapIfId nId |> RoseTree.mapData |> DummyRoot.map
+    | AddCounter pId -> () |> Guid.NewGuid |> addChildCounter |> mapIf (hasId pId) |> RoseTree.map |> DummyRoot.map
+    | Remove nId -> nId |> hasId >> not |> List.filter |> RoseTree.mapChildren |> RoseTree.map |> DummyRoot.map
+    | MoveUp nId -> nId |> swapCounters List.swapWithPrev |> RoseTree.map |> DummyRoot.map
+    | MoveDown nId -> nId |> swapCounters List.swapWithNext |> RoseTree.map |> DummyRoot.map
```

Comment:
I find the point-free style more difficult to read here than the normal "explicit model" style. Unless you strongly disagree, could you revert to "explicit model" style? Then you can also remove a lot of the helper functions, since you can do `{ m with ...}` (I'll leave the naming of `model`/`m` to you). I think this makes the sample easier to understand. I can certainly see and appreciate the aesthetic qualities of consistent point-free style as demonstrated here, but many find it confusing.

Also, note that I'm not a fan of using pipes wherever pipes can be used. For example, I prefer `SomeGlobalState.map not` instead of `not |> SomeGlobalState.map`, and, particularly, `() |> Guid.NewGuid` should be `Guid.NewGuid ()`.

I'll continue reviewing this PR after this is fixed.

Reply:
> Also, note that I'm not a fan of using pipes wherever pipes can be used. For example, I prefer `SomeGlobalState.map not` instead of `not |> SomeGlobalState.map`, and, particularly, `() |> Guid.NewGuid` should be `Guid.NewGuid ()`.

Sure.  I don't mind these changes.  In these small cases, my preference varies.  When writing code though, I strongly prefer to always using pipes.  I find the syntactic consistency helpful when I am in that mindset.  Then when I come back to the code just to read it (as you are), I also dislike some of the piping and remove it accordingly.

> I find the point-free style more difficult to read here than the normal "explicit model" style. Unless you strongly disagree, could you revert to "explicit model" style? Then you can also remove a lot of the helper functions, since you can do `{ m with ...}` (I'll leave the naming of `model`/`m` to you). I think this makes the sample easier to understand. I can certainly see and appreciate the aesthetic qualities of consistent point-free style as demonstrated here, but many find it confusing.

Sure.  I switched to the "explicit model" style and removed the resulting unused helper functions.  I made this change via a particularly minimal syntactic change.  Specifically, I change `msg |> f |> g |> DummyRoot.map` into `{ m with DummyRoot = (msg |> f |> g) m.DummyRoot`.  I am not trying to be difficult or obtuse with this change.  I don't mind considering further alternative ways to express this.

Reply:
Finally home again with access to good tooling. After more carefully reviewing this part of the PR, I actually find I prefer the way it was. I find it elegant. The only things I'd want changed (from your original suggestion) is what I mentioned above:

> For example, I prefer `SomeGlobalState.map not` instead of `not |> SomeGlobalState.map`, and, particularly, `() |> Guid.NewGuid` should be `Guid.NewGuid ()`.

Reply:
Yay!  I am so glad that you like it.

What I find particularly elegant is that the type of the data passed from the left into each pipe operator is always* a function with signature `a -> a`.

*maybe not always, but at least most of them

I recently started writing code like this in my project at work.  I have been on journey into functional programming for about three years now.  The "first" or main idea in functional programming might be to avoid mutation.  The second idea might be to use monads to make mutation avoidance easier.  The third idea might be to use higher-order functions to make mutation avoidance easier.  I don't want to mutate, I studied monads for a long time (and I think I understand them well now), I have been using the given higher-order functions (like `List.map` and `List.filter`) for a long time starting with LINQ, but now I think I have just begun to write my own useful higher-order functions.

Back on topic...

I just pushed a commit that uses the mapping functions again.  I also made the other changes you requested.  In case I didn't do exactly what you wanted, I will leave this open for you to comment again or resolve.